File: src/integrations/clio.ts
// SPDX-License-Identifier: AGPL-3.0-only
// Copyright (C) 2026 Discover Legal
// This program is free software: you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version. See <https://www.gnu.org/licenses/>.

/**
 * Clio practice management integration.
 *
 * Authentication: OAuth 2.0 authorization code grant.
 * Tokens (access + refresh) are persisted to Config.clio.tokensFile.
 * The access token is auto-refreshed whenever it is within 60 seconds of expiry.
 *
 * Region routing: Clio operates four data regions — us, eu, ca, au.
 * Set CLIO_REGION to the firm's region. The base URL is derived from a
 * hard-coded allowlist, so a malformed env var cannot cause SSRF.
 *
 * Security: API key (client secret) never appears in logs; all requests go to
 * hard-coded Clio domains (not configurable); response bodies are capped at 2 MB.
 */

import { readFile, writeFile, mkdir } from "node:fs/promises";
import { dirname } from "node:path";
import { Config } from "../config.js";
import { logger } from "../logger.js";

const CLIO_REGIONS: Record<string, string> = {
  us: "https://app.clio.com",
  eu: "https://eu.app.clio.com",
  ca: "https://ca.app.clio.com",
  au: "https://au.app.clio.com",
};

const RESPONSE_CAP = 2_000_000; // 2 MB

export interface ClioTokens {
  accessToken: string;
  refreshToken: string;
  expiresAt: number;   // unix timestamp in ms
  tokenType: string;
  firmId?: string;
  firmName?: string;
  connectedAt: string; // ISO timestamp
}

export class ClioClient {
  private tokens: ClioTokens | null = null;
  private readonly base: string;

  constructor() {
    const region = Config.clio.region;
    const base = CLIO_REGIONS[region];
    if (!base) throw new Error(`Unknown CLIO_REGION "${region}". Valid values: us, eu, ca, au`);
    this.base = base;
  }

  // ── OAuth helpers ────────────────────────────────────────────────────────

  /** Build the Clio authorization URL to redirect the user to. */
  authUrl(state: string): string {
    const p = new URLSearchParams({
      response_type: "code",
      client_id: Config.clio.clientId,
      redirect_uri: Config.clio.redirectUri,
      state,
    });
    // Include scopes if configured. Clio v4 falls back to the app's portal-level
    // permissions when scope is omitted, so this is optional but recommended for
    // apps that declare explicit scopes in the developer portal.
    if (Config.clio.scopes) p.set("scope", Config.clio.scopes);
    return `${this.base}/oauth/authorize?${p}`;
  }

  /** Exchange an authorization code for tokens and persist them. */
  async exchangeCode(code: string): Promise<void> {
    const res = await fetch(`${this.base}/oauth/token`, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded", Accept: "application/json" },
      body: new URLSearchParams({
        grant_type: "authorization_code",
        code,
        redirect_uri: Config.clio.redirectUri,
        client_id: Config.clio.clientId,
        client_secret: Config.clio.clientSecret,
      }),
    });
    if (!res.ok) throw new Error(`Clio token exchange failed: ${res.status}`);
    const body = await res.json() as { access_token: string; refresh_token: string; expires_in: number; token_type: string };
    this.tokens = {
      accessToken: body.access_token,
      refreshToken: body.refresh_token,
      expiresAt: Date.now() + body.expires_in * 1000,
      tokenType: body.token_type,
      connectedAt: new Date().toISOString(),
    };
    // Fetch firm info to store firmName
    try {
      const me = await this.get("/api/v4/users/who_am_i.json", { fields: "id,name,account{id,name}" }) as { data?: { id?: number; name?: string; account?: { id?: number; name?: string } } };
      if (me?.data?.account) {
        this.tokens.firmId = String(me.data.account.id ?? "");
        this.tokens.firmName = me.data.account.name ?? "";
      }
    } catch { /* non-fatal */ }
    await this.save();
  }

  /** Refresh the access token using the refresh token. */
  private async refresh(): Promise<void> {
    if (!this.tokens) throw new Error("Clio not connected");
    const res = await fetch(`${this.base}/oauth/token`, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded", Accept: "application/json" },
      body: new URLSearchParams({
        grant_type: "refresh_token",
        refresh_token: this.tokens.refreshToken,
        client_id: Config.clio.clientId,
        client_secret: Config.clio.clientSecret,
      }),
    });
    if (!res.ok) {
      // Refresh token invalid — clear stored tokens
      this.tokens = null;
      await this.save();
      throw new Error(`Clio token refresh failed: ${res.status} — reconnect required`);
    }
    const body = await res.json() as { access_token: string; refresh_token?: string; expires_in: number; token_type: string };
    this.tokens = {
      ...this.tokens,
      accessToken: body.access_token,
      refreshToken: body.refresh_token ?? this.tokens.refreshToken,
      expiresAt: Date.now() + body.expires_in * 1000,
      tokenType: body.token_type,
    };
    await this.save();
  }

  /** Return a valid access token, refreshing if within 60s of expiry. */
  private async ensureValid(): Promise<string> {
    if (!this.tokens) throw new Error("Clio not connected — visit /auth/clio/connect");
    if (Date.now() >= this.tokens.expiresAt - 60_000) await this.refresh();
    return this.tokens.accessToken;
  }

  // ── Persistence ──────────────────────────────────────────────────────────

  async load(): Promise<void> {
    try {
      const raw = await readFile(Config.clio.tokensFile, "utf8");
      this.tokens = JSON.parse(raw) as ClioTokens;
    } catch {
      this.tokens = null;
    }
  }

  private async save(): Promise<void> {
    await mkdir(dirname(Config.clio.tokensFile), { recursive: true });
    await writeFile(Config.clio.tokensFile, JSON.stringify(this.tokens ?? null, null, 2), "utf8");
  }

  async disconnect(): Promise<void> {
    this.tokens = null;
    await this.save();
    logger.info("Clio disconnected");
  }

  // ── Status ───────────────────────────────────────────────────────────────

  isConnected(): boolean { return this.tokens !== null; }

  status(): { connected: boolean; firmName?: string; firmId?: string; connectedAt?: string } {
    if (!this.tokens) return { connected: false };
    return { connected: true, firmName: this.tokens.firmName, firmId: this.tokens.firmId, connectedAt: this.tokens.connectedAt };
  }

  // ── HTTP helpers ─────────────────────────────────────────────────────────

  async get(path: string, params?: Record<string, string>): Promise<unknown> {
    const token = await this.ensureValid();
    const url = new URL(path, this.base);
    if (params) for (const [k, v] of Object.entries(params)) url.searchParams.set(k, v);
    const res = await fetch(url, {
      headers: { Authorization: `Bearer ${token}`, Accept: "application/json" },
      signal: AbortSignal.timeout(30_000),
    });
    if (!res.ok) throw new Error(`Clio GET ${path} failed: ${res.status}`);
    const text = await res.text();
    if (text.length > RESPONSE_CAP) throw new Error("Clio response exceeded 2 MB cap");
    return JSON.parse(text);
  }

  async getBuffer(path: string): Promise<Buffer> {
    const token = await this.ensureValid();
    const url = new URL(path, this.base);
    const res = await fetch(url, {
      headers: { Authorization: `Bearer ${token}` },
      signal: AbortSignal.timeout(60_000),
    });
    if (!res.ok) throw new Error(`Clio download ${path} failed: ${res.status}`);
    const ab = await res.arrayBuffer();
    if (ab.byteLength > RESPONSE_CAP) throw new Error("Clio download exceeded 2 MB cap");
    return Buffer.from(ab);
  }

  async post(path: string, body: Record<string, unknown>): Promise<unknown> {
    const token = await this.ensureValid();
    const res = await fetch(new URL(path, this.base), {
      method: "POST",
      headers: { Authorization: `Bearer ${token}`, "Content-Type": "application/json", Accept: "application/json" },
      body: JSON.stringify(body),
      signal: AbortSignal.timeout(30_000),
    });
    if (!res.ok) throw new Error(`Clio POST ${path} failed: ${res.status}`);
    const text = await res.text();
    if (text.length > RESPONSE_CAP) throw new Error("Clio response exceeded 2 MB cap");
    return JSON.parse(text);
  }

  // ── Matter API ───────────────────────────────────────────────────────────

  async listMatters(opts: { status?: string; limit?: number; page?: number } = {}): Promise<unknown> {
    const params: Record<string, string> = {
      fields: "id,display_number,description,status,client{id,name},practice_area{name},open_date,close_date",
      limit: String(opts.limit ?? 50),
      page: String(opts.page ?? 1),
    };
    if (opts.status && opts.status !== "all") params["status[]"] = opts.status;
    return this.get("/api/v4/matters.json", params);
  }

  async getMatter(id: number): Promise<unknown> {
    return this.get(`/api/v4/matters/${id}.json`, {
      fields: "id,display_number,description,status,client{id,name,email_addresses},practice_area{name},open_date,close_date,custom_fields{value,field_name},responsible_attorney{id,name},originating_attorney{id,name}",
    });
  }

  async listDocuments(matterId: number, limit = 50): Promise<unknown> {
    return this.get("/api/v4/documents.json", {
      "matter_id": String(matterId),
      fields: "id,name,content_type,latest_document_version{id,fully_uploaded}",
      limit: String(limit),
    });
  }

  async downloadDocument(documentId: number): Promise<Buffer> {
    return this.getBuffer(`/api/v4/documents/${documentId}/download`);
  }

  async createActivity(matterId: number, opts: { description: string; dateOn: string; durationHours: number; userId?: number }): Promise<unknown> {
    return this.post("/api/v4/activities.json", {
      data: {
        type: "TimeEntry",
        date: opts.dateOn,
        quantity_in_hours: opts.durationHours,
        note: opts.description,
        matter: { id: matterId },
        ...(opts.userId ? { user: { id: opts.userId } } : {}),
      },
    });
  }

  async createNote(matterId: number, subject: string, detail: string): Promise<unknown> {
    return this.post("/api/v4/notes.json", {
      data: {
        subject,
        detail,
        matter: { id: matterId },
      },
    });
  }

  async listContacts(opts: { type?: string; limit?: number } = {}): Promise<unknown> {
    const params: Record<string, string> = {
      fields: "id,name,type,email_addresses,phone_numbers",
      limit: String(opts.limit ?? 50),
    };
    if (opts.type) params["type"] = opts.type;
    return this.get("/api/v4/contacts.json", params);
  }
}

/** Singleton — loaded once at startup, reused across all requests. */
export const clioClient = new ClioClient();

File: biglaw-go/internal/integrations/clio.go
// SPDX-License-Identifier: Apache-2.0
// Copyright (C) 2026 Discover Legal

// Clio practice management integration.
// OAuth 2.0 authorization code grant; tokens persisted to CLIO_TOKENS_FILE.
// Region routing via hard-coded allowlist (SSRF-safe): CLIO_REGION = us|eu|ca|au.
// Response bodies capped at 2 MB.

package integrations

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"sync"
	"time"
)

const clioResponseCap = 2_000_000

var clioRegions = map[string]string{
	"us": "https://app.clio.com",
	"eu": "https://eu.app.clio.com",
	"ca": "https://ca.app.clio.com",
	"au": "https://au.app.clio.com",
}

// ClioTokens holds OAuth2 credentials persisted to disk.
type ClioTokens struct {
	AccessToken  string `json:"accessToken"`
	RefreshToken string `json:"refreshToken"`
	ExpiresAt    int64  `json:"expiresAt"` // unix ms
	TokenType    string `json:"tokenType"`
	FirmID       string `json:"firmId,omitempty"`
	FirmName     string `json:"firmName,omitempty"`
	ConnectedAt  string `json:"connectedAt"`
}

// ClioClient wraps the Clio REST API.
type ClioClient struct {
	mu         sync.Mutex
	tokens     *ClioTokens
	base       string
	tokensFile string
}

var DefaultClioClient *ClioClient

func init() {
	region := os.Getenv("CLIO_REGION")
	if region == "" {
		region = "us"
	}
	base, ok := clioRegions[region]
	if !ok {
		base = clioRegions["us"]
	}
	tokensFile := os.Getenv("CLIO_TOKENS_FILE")
	if tokensFile == "" {
		tokensFile = "./data/clio-tokens.json"
	}
	DefaultClioClient = &ClioClient{base: base, tokensFile: tokensFile}
	_ = DefaultClioClient.Load()
}

// AuthURL returns the OAuth2 authorization URL to redirect the user to.
func (c *ClioClient) AuthURL(state string) string {
	p := url.Values{
		"response_type": {"code"},
		"client_id":     {os.Getenv("CLIO_CLIENT_ID")},
		"redirect_uri":  {os.Getenv("CLIO_REDIRECT_URI")},
		"state":         {state},
	}
	if scopes := os.Getenv("CLIO_SCOPES"); scopes != "" {
		p.Set("scope", scopes)
	}
	return c.base + "/oauth/authorize?" + p.Encode()
}

// ExchangeCode exchanges an authorization code for tokens and persists them.
func (c *ClioClient) ExchangeCode(code string) error {
	form := url.Values{
		"grant_type":    {"authorization_code"},
		"code":          {code},
		"redirect_uri":  {os.Getenv("CLIO_REDIRECT_URI")},
		"client_id":     {os.Getenv("CLIO_CLIENT_ID")},
		"client_secret": {os.Getenv("CLIO_CLIENT_SECRET")},
	}
	data, err := c.oauthPost(form)
	if err != nil {
		return err
	}
	var body struct {
		AccessToken  string `json:"access_token"`
		RefreshToken string `json:"refresh_token"`
		ExpiresIn    int    `json:"expires_in"`
		TokenType    string `json:"token_type"`
	}
	if err := json.Unmarshal(data, &body); err != nil {
		return err
	}
	c.mu.Lock()
	c.tokens = &ClioTokens{
		AccessToken:  body.AccessToken,
		RefreshToken: body.RefreshToken,
		ExpiresAt:    time.Now().UnixMilli() + int64(body.ExpiresIn)*1000,
		TokenType:    body.TokenType,
		ConnectedAt:  time.Now().UTC().Format(time.RFC3339),
	}
	c.mu.Unlock()

	// Try to fetch firm name (non-fatal)
	if me, err := c.Get("/api/v4/users/who_am_i.json", map[string]string{
		"fields": "id,name,account{id,name}",
	}); err == nil {
		var meResp struct {
			Data struct {
				Account struct {
					ID   int    `json:"id"`
					Name string `json:"name"`
				} `json:"account"`
			} `json:"data"`
		}
		if b, err := json.Marshal(me); err == nil {
			_ = json.Unmarshal(b, &meResp)
			c.mu.Lock()
			c.tokens.FirmID = fmt.Sprintf("%d", meResp.Data.Account.ID)
			c.tokens.FirmName = meResp.Data.Account.Name
			c.mu.Unlock()
		}
	}
	return c.save()
}

func (c *ClioClient) refresh() error {
	c.mu.Lock()
	if c.tokens == nil {
		c.mu.Unlock()
		return fmt.Errorf("Clio not connected")
	}
	rt := c.tokens.RefreshToken
	c.mu.Unlock()

	form := url.Values{
		"grant_type":    {"refresh_token"},
		"refresh_token": {rt},
		"client_id":     {os.Getenv("CLIO_CLIENT_ID")},
		"client_secret": {os.Getenv("CLIO_CLIENT_SECRET")},
	}
	data, err := c.oauthPost(form)
	if err != nil {
		c.mu.Lock()
		c.tokens = nil
		c.mu.Unlock()
		_ = c.save()
		return fmt.Errorf("Clio token refresh failed: %w — reconnect required", err)
	}
	var body struct {
		AccessToken  string `json:"access_token"`
		RefreshToken string `json:"refresh_token"`
		ExpiresIn    int    `json:"expires_in"`
		TokenType    string `json:"token_type"`
	}
	if err := json.Unmarshal(data, &body); err != nil {
		return err
	}
	c.mu.Lock()
	c.tokens.AccessToken = body.AccessToken
	if body.RefreshToken != "" {
		c.tokens.RefreshToken = body.RefreshToken
	}
	c.tokens.ExpiresAt = time.Now().UnixMilli() + int64(body.ExpiresIn)*1000
	c.tokens.TokenType = body.TokenType
	c.mu.Unlock()
	return c.save()
}

func (c *ClioClient) ensureValid() (string, error) {
	c.mu.Lock()
	if c.tokens == nil {
		c.mu.Unlock()
		return "", fmt.Errorf("Clio not connected — visit /auth/clio/connect")
	}
	exp := c.tokens.ExpiresAt
	at := c.tokens.AccessToken
	c.mu.Unlock()

	if time.Now().UnixMilli() >= exp-60_000 {
		if err := c.refresh(); err != nil {
			return "", err
		}
		c.mu.Lock()
		at = c.tokens.AccessToken
		c.mu.Unlock()
	}
	return at, nil
}

// Load reads persisted tokens from disk.
func (c *ClioClient) Load() error {
	data, err := os.ReadFile(c.tokensFile)
	if err != nil {
		return err
	}
	var t ClioTokens
	if err := json.Unmarshal(data, &t); err != nil {
		return err
	}
	c.mu.Lock()
	c.tokens = &t
	c.mu.Unlock()
	return nil
}

func (c *ClioClient) save() error {
	c.mu.Lock()
	data, err := json.MarshalIndent(c.tokens, "", "  ")
	c.mu.Unlock()
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(c.tokensFile), 0755); err != nil {
		return err
	}
	return os.WriteFile(c.tokensFile, data, 0600)
}

// Disconnect clears stored tokens.
func (c *ClioClient) Disconnect() error {
	c.mu.Lock()
	c.tokens = nil
	c.mu.Unlock()
	return c.save()
}

// IsConnected returns true when tokens are available.
func (c *ClioClient) IsConnected() bool {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.tokens != nil
}

// IsConfigured reports whether the Clio OAuth app credentials are present.
// Mirrors Config.clio.enabled in the TS implementation (CLIO_CLIENT_ID set).
func (c *ClioClient) IsConfigured() bool {
	return os.Getenv("CLIO_CLIENT_ID") != ""
}

// Status returns the connection status.
func (c *ClioClient) Status() map[string]interface{} {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.tokens == nil {
		return map[string]interface{}{"connected": false}
	}
	return map[string]interface{}{
		"connected":   true,
		"firmName":    c.tokens.FirmName,
		"firmId":      c.tokens.FirmID,
		"connectedAt": c.tokens.ConnectedAt,
	}
}

// Get performs a Clio API GET request.
func (c *ClioClient) Get(path string, params map[string]string) (interface{}, error) {
	token, err := c.ensureValid()
	if err != nil {
		return nil, err
	}
	u, err := url.Parse(c.base + path)
	if err != nil {
		return nil, err
	}
	q := u.Query()
	for k, v := range params {
		q.Set(k, v)
	}
	u.RawQuery = q.Encode()

	client := &http.Client{Timeout: 30 * time.Second}
	req, err := http.NewRequest("GET", u.String(), nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Accept", "application/json")
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return nil, fmt.Errorf("Clio GET %s: HTTP %d", path, resp.StatusCode)
	}
	data, err := io.ReadAll(io.LimitReader(resp.Body, clioResponseCap))
	if err != nil {
		return nil, err
	}
	var result interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, err
	}
	return result, nil
}

// ListMatters returns matters filtered by status.
func (c *ClioClient) ListMatters(status string, limit, page int) (interface{}, error) {
	params := map[string]string{
		"fields": "id,display_number,description,status,client{id,name},practice_area{name},open_date,close_date",
		"limit":  fmt.Sprintf("%d", limit),
		"page":   fmt.Sprintf("%d", page),
	}
	if status != "" && status != "all" {
		params["status[]"] = status
	}
	return c.Get("/api/v4/matters.json", params)
}

// GetMatter returns full matter details including client, practice area,
// responsible attorney, and custom fields.
func (c *ClioClient) GetMatter(id int) (interface{}, error) {
	return c.Get(fmt.Sprintf("/api/v4/matters/%d.json", id), map[string]string{
		"fields": "id,display_number,description,status,client{id,name,email_addresses},practice_area{name},open_date,close_date,custom_fields{value,field_name},responsible_attorney{id,name},originating_attorney{id,name}",
	})
}

// ListDocuments lists documents for a matter.
func (c *ClioClient) ListDocuments(matterID int, limit int) (interface{}, error) {
	return c.Get("/api/v4/documents.json", map[string]string{
		"matter_id": fmt.Sprintf("%d", matterID),
		"fields":    "id,name,content_type,latest_document_version{id,fully_uploaded}",
		"limit":     fmt.Sprintf("%d", limit),
	})
}

// DownloadDocument fetches the raw bytes of a document. Clio serves the
// content via a redirect to blob storage, which the default client follows
// (the Authorization header is dropped on cross-host redirects). The body is
// capped at 2 MB, matching the TS getBuffer().
func (c *ClioClient) DownloadDocument(documentID int) ([]byte, error) {
	token, err := c.ensureValid()
	if err != nil {
		return nil, err
	}
	path := fmt.Sprintf("/api/v4/documents/%d/download", documentID)
	client := &http.Client{Timeout: 60 * time.Second}
	req, err := http.NewRequest("GET", c.base+path, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+token)
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return nil, fmt.Errorf("Clio download %s: HTTP %d", path, resp.StatusCode)
	}
	data, err := io.ReadAll(io.LimitReader(resp.Body, clioResponseCap+1))
	if err != nil {
		return nil, err
	}
	if len(data) > clioResponseCap {
		return nil, fmt.Errorf("Clio download exceeded 2 MB cap")
	}
	return data, nil
}

// CreateActivity logs a time entry in Clio.
func (c *ClioClient) CreateActivity(matterID int, description, dateOn string, durationHours float64) (interface{}, error) {
	body := map[string]interface{}{
		"data": map[string]interface{}{
			"type":              "TimeEntry",
			"date":              dateOn,
			"quantity_in_hours": durationHours,
			"note":              description,
			"matter":            map[string]int{"id": matterID},
		},
	}
	return c.post("/api/v4/activities.json", body)
}

// CreateNote posts a note to a matter — used to save synthesis output and
// research memos back into the client file.
func (c *ClioClient) CreateNote(matterID int, subject, detail string) (interface{}, error) {
	body := map[string]interface{}{
		"data": map[string]interface{}{
			"subject": subject,
			"detail":  detail,
			"matter":  map[string]int{"id": matterID},
		},
	}
	return c.post("/api/v4/notes.json", body)
}

// ListContacts lists contacts, optionally filtered by type (Person or Company).
func (c *ClioClient) ListContacts(contactType string, limit int) (interface{}, error) {
	params := map[string]string{
		"fields": "id,name,type,email_addresses,phone_numbers",
		"limit":  fmt.Sprintf("%d", limit),
	}
	if contactType != "" {
		params["type"] = contactType
	}
	return c.Get("/api/v4/contacts.json", params)
}

func (c *ClioClient) post(path string, body interface{}) (interface{}, error) {
	token, err := c.ensureValid()
	if err != nil {
		return nil, err
	}
	b, err := json.Marshal(body)
	if err != nil {
		return nil, err
	}
	client := &http.Client{Timeout: 30 * time.Second}
	req, err := http.NewRequest("POST", c.base+path, io.NopCloser(
		func() io.Reader {
			var r io.Reader = &bytesReader{b: b}
			return r
		}(),
	))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return nil, fmt.Errorf("Clio POST %s: HTTP %d", path, resp.StatusCode)
	}
	data, err := io.ReadAll(io.LimitReader(resp.Body, clioResponseCap))
	if err != nil {
		return nil, err
	}
	var result interface{}
	_ = json.Unmarshal(data, &result)
	return result, nil
}

func (c *ClioClient) oauthPost(form url.Values) ([]byte, error) {
	client := &http.Client{Timeout: 15 * time.Second}
	req, err := http.NewRequest("POST", c.base+"/oauth/token",
		io.NopCloser(func() io.Reader {
			return &stringReader{s: form.Encode()}
		}()))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/json")
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return nil, fmt.Errorf("Clio OAuth: HTTP %d", resp.StatusCode)
	}
	return io.ReadAll(io.LimitReader(resp.Body, clioResponseCap))
}

// minimal io.Reader wrappers to avoid "bytes" import cycle
type bytesReader struct {
	b   []byte
	pos int
}

func (r *bytesReader) Read(p []byte) (int, error) {
	if r.pos >= len(r.b) {
		return 0, io.EOF
	}
	n := copy(p, r.b[r.pos:])
	r.pos += n
	return n, nil
}

type stringReader struct {
	s   string
	pos int
}

func (r *stringReader) Read(p []byte) (int, error) {
	if r.pos >= len(r.s) {
		return 0, io.EOF
	}
	n := copy(p, r.s[r.pos:])
	r.pos += n
	return n, nil
}
